### ISSUE TITLE
Identify property by issuance transaction

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/spec#280]
The current id scheme with ascending numbers based on the complete history is _extremely_ fragile.

Think of the two cases:
1. A client parses all transactions and there is one false interpretation of a simple send, e.g. due to a corrupt file, ... in the worst case the send is not recognized as such and a **single branch** of the whole system state is changed / black / invalid.
2. A client parses all transactions and there is one false interpretation of a property creation, e.g. due to a corrupt file, ... in the worst case **all following** transactions are completely changed / black / invalid.

I suggest to support identification of properties by some other mechanism which could be based on the (partial) hash of the creation transaction. I'm not sure, if this is the best approach, but it would be an improvement.

As bonus on top the property identifier would be known before the issuance transaction confirms and this enables chaining of [Tx 54: New Property with Managed Number of Tokens](https://github.com/mastercoin-MSC/spec#new-property-with-managed-number-of-tokens) and [Tx 55: Granting Tokens for a Managed Property](https://github.com/mastercoin-MSC/spec#granting-tokens-for-a-managed-property). This was sort of discussed in an Omni thread: https://github.com/mastercoin-MSC/omniwallet/issues/1062#issuecomment-60933928.

If the scope and context of an application is isolated, then a client may not even be required to parse all Mastercoin transactions, but only those after and related to the issuance of a property.

**Comment by zathras-crypto** [comment]
Playing devil's advocate here but why introduce that level of complexity in the first place (wrt property creation)?  Why not simply introduce a new version of the property creation transactions that includes a  field for the requested propID.  If it's already in use when the TX is parsed it's invalid.

For example I issue a tx creating a property and use '3' as the field for the requested property ID.  This tx is invalid and the property doesn't get created because the property ID is already in use.

Alternatively, let's use the example that I'm the popular 888 Casino website and I want to shift to my own crypto-backed tokens for some purpose - I issue a tx creating a property and use '888' as the field for the requested property ID.  This tx is valid and the property gets created because the property ID is available.

There would be some (not dramatic) coding changes in Core yep but this seems like a fairly simplistic model spec & code wise to move away from sequential prop IDs (which have various problems I won't re-iterate here).

Thanks
Z

**Comment by dexX7** [comment]
> Why not simply introduce a new version of the property creation transactions that includes a field for the requested propID. If it's already in use when the TX is parsed it's invalid.

Unless I miss something, this doesn't seem to solve the issue at all, because "full knowledge" (= every Master transaction processed correctly) is still required to determine, whether a property identifier is valid.

There are actually two seperate aspects:

1. full block chain / knowledge about transactions is required
2. complete and correct knowledge about meta transactions

What I proposed earlier, using RIPEMD-160(transaction id) or a "partial hash piece" as identifier, clears the second point, but I just noticed full block chain knowledge is still required. To solve both, only full transaction hashes seems to be viable.

> level of complexity

To what are you referring here? The base idea is to use transaction ids to identify properties - which seems like a straightforward approach, but I'm biased against raw hashes due to the length.

OTOH there are several parts where the transaction size could be reduced significantly (e.g. using multisig output for change), but it doesn't seem to be an issue at this point.

Right now a "standard" simple send has a size of about 340 byte and using hashes instead of property ids comes with an overhead of 28 byte (32 byte hash vs. 4 byte property id) -- not very much imho. 

Bitcoin Core 0.9 has "static fees" of x BTC/1000 byte, so it's "for free", because there is quite some space left. With 0.10 it's going to be fees paid per byte, and assuming 0.0001 BTC/kB, the overhead would equal additional costs of 0.0000028 BTC, less than 0.01 of a penny at the moment.

Let me be clear, I'm biased against those 28 bytes, but in the broader picture even a slightly improved input/change handling would have greater consequences - not even speaking of actually using "low output values" on mainnet, which seems to be the most significant, but unnecessary, cost factor at the moment.

**Comment by zathras-crypto** [comment]
> Unless I miss something, this doesn't seem to solve the issue at all, because "full knowledge" (= every Master transaction processed correctly) is still required to determine, whether a property identifier is valid.

It's not exactly thoroughly thought out :P but I thought of this as only requiring us to know about previous creation transactions, not "full knowledge" as you put it of everything - create-property transaction validity is not dependent on anything other than previous prop creation's (assuming we allowed it to be specified) - I may be wrong though will need to think it through a little more.

> To what are you referring here?

Simple change/scope control - allowing a property ID to be specified doesn't involve a large degree of change, but on initially reading your proposed move to property hashes for identification instead of property ID's seemed like a substantially 'bigger' change.  

Also just for 2 cents - does your method not require knowledge of previous transactions also?  Utilizing the hash is atomic sure but we still need previous transactions to code in collision handling (no matter how unlikely we would need to cater for it I would assume?).

Thanks :)
Z

**Comment by dexX7** [comment]
> create-property transaction validity is not dependent on anything other than previous prop creation's

The dependency on previous property creations *is* the issue I see. ;) And this seems to be a similar problem with user defined ids as well as ascending auto-ids. You're right though, not every Master transaction is required to be processed correctly, but only those involving property creation.

> collision handling

Let's stick to transaction hashes as identifier. In this case the property id == transaction id of creation transaction. One would be able to fetch the transaction and even without any other transaction the property information would be available.

The txid of a Bitcoin transaction is a SHA256(SHA256(raw_transaction)) hash, and I think it's safe to assume collision resistance is given.

**Comment by ABISprotocol** [comment]
History forgotten

**Comment by zathras-crypto** [comment]
> The txid of a Bitcoin transaction is a SHA256(SHA256(raw_transaction)) hash, and I think it's safe to assume collision resistance is given.

Fair point :)

> Let's stick to transaction hashes as identifier.

This is a pretty fundamental change in the code (switching property ID's from uint32 to uint256) but could be done.  Before we think about code though let's get some more views/feedback in.

Note, transaction malleability could be used to play havoc with anything that involves a property creation and zero conf if we did go this route, so would need to tread carefully for pending/unconfirmed perhaps.

**Comment by zathras-crypto** [comment]
Ah - think I see an issue - we couldn't treat all creations as atomic because 51 requires prior knowledge (only allowed one active crowdsale).

**Comment by dexX7** [comment]
Good point. It still narrows the scope of "all 'create crowdsale' transactions" down to "all 'create crowdsale' transactions of address x".

Or, as alternative and further protocol change, multiple crowdsales per address are allowed and crowdsale participation becomes a more active process where the crowdsale-to-participate-in is explicitly selected. This would be in line with #277 ("I unintentionally participated in a crowdsale") and #261 ("Allow users to specify the specific terms of an offer they are responding to").

> transaction malleability

That's unfortunally true. Depending one the use-case one could chain transactions where either the whole chain is valid or invalidated due to a altered crowdsale transaction (assuming the issuer wants to do something sneaky with unconfirmed transactions for some reason).

The overall goal is not necessarily to get down to robustness on a single transaction basis, but to prevent brutal butterfly effects, so to speak. (e.g. all property ids off by one number)

> This is a pretty fundamental change in the code

Yeah, probably. ;) I just focused on the theory and haven't thought much about potential implementations. The most annoying part is probably to support both schemes at the same time (to process the "old" crowdsales), but this an issue in general and needs to be addressed with every change.
